### PR TITLE
Request to add gun operations

### LINKED ISSUE #17: Gun
- [ ]  Magazine (include ammo)
- [ ]  Turret 
  * Should include magazine
  * There must be a method reload()
- [ ]  Railgun (can shoot all magnitable item)

### PR DESCRIPTION
**Pull request by ghost**
Gun operations dab41f8
- Railgun now can shoot IMagnitable items
- Turret now can shoot
- Add 2 new bullet classes: ArmorBullet and UsualBullet

Resolve #17

### CODE REVIEW

**Comment by ViRGiL175** [comment]
> Add 2 new bullet classes: ArmorBullet and UsualBullet

`AverageBullet` is better choice as for me.

**Comment by ViRGiL175** [comment]
https://github.com/ViRGiL175/java-git-starship/blob/dab41f85387ff6859afb937c36bb7188f99fe7b0/src/main/java/rape/brutal/gitstarship/ammunition/Gun.java#L53

You have abstract class, but there is unabstract method returning pointless zero. Make it abstract too.

File: src/main/java/rape/brutal/gitstarship/ammunition/Railgun.java [hunk 1]
```diff
@@ -5,14 +5,26 @@
...
+
 /**
  * Created by haze on 22.03.2017.
  */
 public class Railgun extends Gun {
 
-    private int cost;
-    private String name;
-    private int size;
+
```
Comment (ViRGiL175) [inline src/main/java/rape/brutal/gitstarship/ammunition/Railgun.java]: Redundant empty line.

File: src/main/java/rape/brutal/gitstarship/ammunition/Railgun.java [hunk 2]
```diff
@@ -5,14 +5,26 @@
...
+    public Railgun(int damage, String name, int size, int cost) {
+        super(damage, name, size, cost);
+    }
+
+
```
Comment (ViRGiL175) [inline src/main/java/rape/brutal/gitstarship/ammunition/Railgun.java]: Redundant empty line.

File: src/main/java/rape/brutal/gitstarship/ammunition/Gun.java [hunk 3]
```diff
@@ -15,6 +15,14 @@
     protected int damage;
     protected String name;
     protected int size;
+    protected int cost;
+
+    public Gun(int damage, String name, int size, int cost) {
```
Comment (ViRGiL175) [inline src/main/java/rape/brutal/gitstarship/ammunition/Gun.java:20]: Why is there a damage? Damage gets from ammo, is not it? Or damage gets from ammo and from gun?

File: src/main/java/rape/brutal/gitstarship/ammunition/Magazine.java [hunk 4]
```diff
@@ -34,8 +43,19 @@ public void setAmmo(Ammo[] ammo) {
         this.ammo = ammo;
     }
 
-    public void removeAmmo() {
-
+    public int removeAmmo() {
```
Comment (ViRGiL175) [inline src/main/java/rape/brutal/gitstarship/ammunition/Magazine.java]: What kind of value this method returns?

File: src/main/java/rape/brutal/gitstarship/ammunition/Magazine.java [hunk 5]
```diff
@@ -34,8 +43,19 @@ public void setAmmo(Ammo[] ammo) {
...
+        for (int i = 0; i < 30; i++) {
```
Comment (ViRGiL175) [inline src/main/java/rape/brutal/gitstarship/ammunition/Magazine.java]: IDEA syntax analyzer shows error here.

File: src/main/java/rape/brutal/gitstarship/ammunition/Magazine.java [hunk 6]
```diff
@@ -34,8 +43,19 @@ public void setAmmo(Ammo[] ammo) {
...
+            if (this.ammo[i] != null) {
+                this.ammo[i] = null;
+                return this.ammo[i].getDamage();
+            }
+            if (this.ammo[29] == null) {
```
Comment (ViRGiL175) [inline src/main/java/rape/brutal/gitstarship/ammunition/Magazine.java]: And if there are other bullets, except for the 29th?

File: src/main/java/rape/brutal/gitstarship/ammunition/Magazine.java [hunk 7]
```diff
@@ -6,17 +6,26 @@
...
+    protected String name;
+    protected int size;
+    protected int cost;
+
+    public Magazine(int capacity, String name, int size, int cost) {
+        this.capacity = capacity;
+        this.name = name;
+        this.size = size;
+        this.cost = cost;
+        this.ammo = new UsualBullet[30];
```
Comment (ViRGiL175) [inline src/main/java/rape/brutal/gitstarship/ammunition/Magazine.java]: Why exact `UsualBullet`? Why not `ArmorBullet`? How I can load `ArmorBullet`s in `Magazine`?

File: src/main/java/rape/brutal/gitstarship/ammunition/Magazine.java [hunk 8]
```diff
@@ -6,17 +6,26 @@
...
-    private String name;
-    private int size;
-    private int cost;
+    protected int capacity;
+    protected Ammo[] ammo;
+    protected String name;
+    protected int size;
+    protected int cost;
+
+    public Magazine(int capacity, String name, int size, int cost) {
```
Comment (ViRGiL175) [inline src/main/java/rape/brutal/gitstarship/ammunition/Magazine.java]: You set capacity here, but don't use it with array initialization.

Array always have 30 bullets, but you can have different capacity.

Comment (ViRGiL175) [reply src/main/java/rape/brutal/gitstarship/ammunition/Magazine.java]: Also you can specify name, size and cost right in constructor.

File: src/main/java/rape/brutal/gitstarship/ammunition/Railgun.java [hunk 9]
```diff
@@ -5,14 +5,26 @@
...
 /**
  * Created by haze on 22.03.2017.
  */
 public class Railgun extends Gun {
 
-    private int cost;
-    private String name;
-    private int size;
+
+    public Railgun(int damage, String name, int size, int cost) {
```
Comment (ViRGiL175) [inline src/main/java/rape/brutal/gitstarship/ammunition/Railgun.java]: You can specify name, size and cost right in constructor.

File: src/main/java/rape/brutal/gitstarship/ammunition/Railgun.java [hunk 10]
```diff
@@ -5,14 +5,26 @@
...
+        super(damage, name, size, cost);
+    }
+
+
+    public int fire(IStorageable iStorageable) {
+        if (iStorageable instanceof IMagnetable) {
+            return damage;
```
Comment (ViRGiL175) [inline src/main/java/rape/brutal/gitstarship/ammunition/Railgun.java]: Wrong. Damage should vary rely to IMagnetable object parameters. From size, for example.

File: src/main/java/rape/brutal/gitstarship/ammunition/Turret.java [hunk 11]
```diff
@@ -10,12 +10,15 @@
  */
 public class Turret extends Gun {
 
-    private int cost;
-    private String name;
-    private int size;
 
     private Magazine loadedMagazine;
 
+    public Turret(int damage, String name, int size, int cost, Magazine magazine) {
```
Comment (ViRGiL175) [inline src/main/java/rape/brutal/gitstarship/ammunition/Turret.java]: You can specify name, size and cost right in constructor.

File: src/main/java/rape/brutal/gitstarship/ammunition/Turret.java [hunk 12]
```diff
@@ -25,7 +28,11 @@ public void setLoadedMagazine(Magazine loadedMagazine) {
     }
 
     public void reload(Magazine magazine) {
+        this.loadedMagazine = magazine;
+    }
 
+    public int fire() {
+        return this.loadedMagazine.removeAmmo();
```
Comment (ViRGiL175) [inline src/main/java/rape/brutal/gitstarship/ammunition/Turret.java:33]: Unclear coding. Damage returns from ammo removing? Absurd.

**Review by ViRGiL175 — CHANGES_REQUESTED** [review]
VERYBAD!!! Worst request ever.

**Comment by ViRGiL175** [comment]
Reviewed.

**Comment by ghost** [comment]
I made a fixes and posted the changes using amend. Please, reviewed again!

File: src/main/java/rape/brutal/gitstarship/ammunition/Magazine.java [hunk 13]
```diff
@@ -30,12 +39,26 @@ public void setCapacity(int capacity) {
...
 
-    public void setAmmo(Ammo[] ammo) {
-        this.ammo = ammo;
+    public void setAmmo(int numberOfBullet, Ammo bullet) {
+        this.ammo[numberOfBullet] = bullet;
     }
 
-    public void removeAmmo() {
+    public int removeAmmo() {
+        int i = 0;
```
Comment (ViRGiL175) [inline src/main/java/rape/brutal/gitstarship/ammunition/Magazine.java:47]: Unused variable.

File: src/main/java/rape/brutal/gitstarship/ammunition/Magazine.java [hunk 14]
```diff
@@ -30,12 +39,26 @@ public void setCapacity(int capacity) {
...
     }
 
-    public void setAmmo(Ammo[] ammo) {
-        this.ammo = ammo;
+    public void setAmmo(int numberOfBullet, Ammo bullet) {
+        this.ammo[numberOfBullet] = bullet;
     }
 
-    public void removeAmmo() {
+    public int removeAmmo() {
```
Comment (ViRGiL175) [inline src/main/java/rape/brutal/gitstarship/ammunition/Magazine.java:46]: Why is `removeAmmo` method calculates damage? Absurd.

File: src/main/java/rape/brutal/gitstarship/ammunition/Railgun.java [hunk 15]
```diff
@@ -5,21 +5,37 @@
...
-    private String name;
-    private int size;
+    public Railgun() {
+        super(1000, "Railgun", 10, 60);
+    }
+
+    public int fire(IStorageable iStorageable) {
+        if (iStorageable instanceof IMagnetable) {
+            System.out.println("Раилган пальнул!");
+            return iStorageable.getSize() * 10;
```
Comment (ViRGiL175) [inline src/main/java/rape/brutal/gitstarship/ammunition/Railgun.java:23]: Is IStorageable removed from storage?

File: src/main/java/rape/brutal/gitstarship/ammunition/Railgun.java [hunk 16]
```diff
@@ -5,21 +5,37 @@
...
 public class Railgun extends Gun {
 
-    private int cost;
-    private String name;
-    private int size;
+    public Railgun() {
+        super(1000, "Railgun", 10, 60);
+    }
+
+    public int fire(IStorageable iStorageable) {
```
Comment (ViRGiL175) [inline src/main/java/rape/brutal/gitstarship/ammunition/Railgun.java:20]: Overloaded method breaks interface. You should use interface method.

File: src/main/java/rape/brutal/gitstarship/ammunition/Railgun.java [hunk 17]
```diff
@@ -5,21 +5,37 @@
...
+        return 0;
+    }
 
     @Override
     public int getSize() {
         return this.size;
     }
 
     @Override
+    public int fire() {
```
Comment (ViRGiL175) [inline src/main/java/rape/brutal/gitstarship/ammunition/Railgun.java:34]: This method should be used.

**Review by ViRGiL175 — CHANGES_REQUESTED** [review]
Bad.

**Comment by ViRGiL175** [comment]
Reviewed.

**Pull request closed without merging**